### PR TITLE
Feature handle dropped 30

### LINKED ISSUE #30: log and handle dropped packets to send
> meshlink_send / channel_send / c->utcp->send currently does not log when packets are getting dropped and in utcp the meshlink_send / channel_send return value for success is ignored, it happily continues to send on ack although it's clear it'll cause a retransmit some time
> https://github.com/meshlink/utcp/blob/master/utcp.c#L604
> https://github.com/meshlink/meshlink/blob/125279ed468788f67d4a6dc64353382bc88b8a6d/src/meshlink.c#L2317
> https://github.com/meshlink/meshlink/blob/125279ed468788f67d4a6dc64353382bc88b8a6d/src/meshlink.c#L1228
> 
> in utcp.c there are other calls to utcp->send too that at least should log when there's a send failure too

### PR DESCRIPTION
**Pull request by everHannes**
Fixes #30
based on #35
linked with utcp https://github.com/meshlink/utcp/pull/12

### CODE REVIEW

**Comment by everHannes** [comment]
works now, along with 8a6c55c I had to clean and rebuild meshlink
there must have been some compiled leftover breaking things
possibly due to the callback error reverted in a3653fa

**Comment by gsliepen** [comment]
On Tue, Sep 06, 2016 at 04:13:58AM -0700, everHannes wrote:

> Assigned #36 to @gsliepen.

Hm, I still think this is the wrong approach. Sure, detecting send
errors and not blindly continuing to do useless sends is good. But I
have strong doubts about adding a queue of packets that should've been
sent. You ARE creating another buffer this way, and while the 1 ms
timeout will cause it to empty quickly and perhaps deal with short
glitches on wireless networks, this may cause bufferbloat on links that
really are congested and that might suddenly have a big latency? If
SO_SNDBUF is full, this will cause new outgoing packets to be queued. If
it takes longer than expected to clear the queue, the retransmit timers
will also kick in and send those packets again right after the queue is
empty. At least, that's what I think will happen.

There are also two scenarios; either MeshLink is communicating via UDP,
in which case UTCP's retransmission timers should deal with dropped
packets, or MeshLink is communicating via TCP. In the latter case, you
can be 99,99% sure any packet sent will eventually arrive. Maybe this
can be signaled to UTCP, in which case it can greatly increase the
retransmit timer.

In both cases, failure to send a packet because the kernel's send buffer
is full should cause UTCP to reduce the congestion window, and
immediately stop sending more packets if it was doing so from a loop of
course (as you already do).

You mentioned in other tickets that you checked that it worked, what
about this branch? What differences did you see?

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

**Comment by everHannes** [comment]
> You ARE creating another buffer this way, and while the 1 ms timeout will cause it to empty quickly and perhaps deal with short glitches on wireless networks, this may cause bufferbloat on links that really are congested and that might suddenly have a big latency?

you are messing the pull requests here, better view it on github ;-)
the 1ms timout only applies to #33 and I already fixed it to only come into affect in one special case that is when the signal pipe keeps being triggered but the sndbuf is full
and that double queue already existed before. I didn't completely fix this but at least utcp ack doesn't mess with it anymore, see #41

**Comment by everHannes** [comment]
> You mentioned in other tickets that you checked that it worked, what about this branch? What differences did you see?

well being short on time I didn't do much testing to it yet, just tested it builds and network communication still works

**Comment by gsliepen** [comment]
On Tue, Sep 06, 2016 at 08:26:07AM -0700, everHannes wrote:

> > You ARE creating another buffer this way, and while the 1 ms timeout will cause it to empty quickly and perhaps deal with short glitches on wireless networks, this may cause bufferbloat on links that really are congested and that might suddenly have a big latency?
> 
> you are messing the pull requests here, better view it on github ;-)

I did mess it up, but even on github it's sometimes hard to see what is
part of the topic when the pull requests depends on other pull requests
:)

> the 1ms timout only applies to #33 and I already fixed it to only come into affect in one special case that is when the signal pipe keeps being triggered but the sndbuf is full
> and that double queue already existed before. I didn't completely fix this but at least utcp ack doesn't mess with it anymore, see #41 

Indeed. Well, I don't see it crash here, but on the other hand, I don't
see a noticable improvement in throughput either. I only have one
Windows machine though, the rest is Linux. Can you tell me more about
your test setup? Is it wireless or wired, and which OSes are being used?
Any on VMs? Sorry I'm not being helpful.

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

File: src/sptps.h [hunk 1]
```diff
@@ -46,7 +46,7 @@
 #define SPTPS_MTU (PAYLOAD_MTU - 19)
 #define SPTPS_DATAGRAM_MTU (PAYLOAD_MTU - 21)
 
-typedef bool (*send_data_t)(void *handle, uint8_t type, const void *data, size_t len);
+typedef int (*send_data_t)(void *handle, uint8_t type, const void *data, size_t len);
```
Comment (a-krebs) [inline src/sptps.h:52]: The desired return values for this function signature should be documented.

File: src/sptps.h [hunk 2]
```diff
@@ -90,7 +90,7 @@ extern void sptps_log_stderr(sptps_t *s, int s_errno, const char *format, va_lis
 extern void (*sptps_log)(sptps_t *s, int s_errno, const char *format, va_list ap);
 extern bool sptps_start(sptps_t *s, void *handle, bool initiator, bool datagram, ecdsa_t *mykey, ecdsa_t *hiskey, const char *label, size_t labellen, send_data_t send_data, receive_record_t receive_record);
 extern bool sptps_stop(sptps_t *s);
-extern bool sptps_send_record(sptps_t *s, uint8_t type, const void *data, uint16_t len);
+extern int sptps_send_record(sptps_t *s, uint8_t type, const void *data, uint16_t len);
```
Comment (a-krebs) [inline src/sptps.h:97]: The desired return values for this function signature should be documented.

File: src/meshlink.c [hunk 3]
```diff
@@ -556,25 +556,25 @@ static bool finalize_join(meshlink_handle_t *mesh) {
...
-    return true;
+    return 0;
 }
 
 static bool invitation_receive(void *handle, uint8_t type, const void *msg, uint16_t len) {
     meshlink_handle_t* mesh = handle;
     switch(type) {
         case SPTPS_HANDSHAKE:
-            return sptps_send_record(&(mesh->sptps), 0, mesh->cookie, sizeof mesh->cookie);
+            return !sptps_send_record(&(mesh->sptps), 0, mesh->cookie, sizeof mesh->cookie);
```
Comment (a-krebs) [inline src/meshlink.c:580]: Oh, this is conversion from `int` to `bool`, correct? Maybe add a short comment to make this more obvious.

Comment (everHannes) [reply src/meshlink.c:580]: Yes and I agree there's lots to be documented. It all just returns the send errno if set or -1 when some other error occured, 0 on success.

**Comment by everHannes** [comment]
> I did mess it up, but even on github it's sometimes hard to see what is
> part of the topic when the pull requests depends on other pull requests
> :)

yeah, that's not so great I agree
what I commonly do is open the branches view and click on compare then select the branches or commits to compare, so e.g.
https://github.com/meshlink/meshlink/compare/feature/fix_o_nonblock...feature/handle_dropped_30

> Indeed. Well, I don't see it crash here, but on the other hand, I don't
> see a noticable improvement in throughput either. I only have one
> Windows machine though, the rest is Linux. Can you tell me more about
> your test setup? Is it wireless or wired, and which OSes are being used?
> Any on VMs? Sorry I'm not being helpful.

Our test setup (that is one hosteurope driven domain server and two laptop i5 clients, wired and all windows) didn't show any improvement yet either. But it showed we're currently facing completely different issues from this, rather related to disk access.
The changes I implemented mostly were meant to improve on congestion and personnally I got the impression of at least some improvement with my virtual machine setup. But would need further testing to verify.
Being short on time I just went ahead and implemented what came to mind analysing the code. When you look up the tickets and comments in code I think I found good argue to most of the changes.
In practice testing however is a completely different story I agree.

**Pull request merged**